### DISCUSSION TITLE
[Proposal] Post-Fix Operator for Flip-Flopping Booleans

### DISCUSSION THREAD

**Discussion by scott-parkhill** [dotnet/csharplang#4632] [Language Ideas]
This is a simple proposal. When you have a boolean property or variable and you want to flip-flop between true and false values, you have to write `some_bool = !some_bool;`, or otherwise specify `true` or `false`. It's a simple thing, but it'd be nice to be able to just write `!some_bool!;`, which would be equivalent to `some_bool = !some_bool;`.

It would be nice too if this acted as a post-fix operator, so that if `some_bool` is `true`, then after the execution of the following if-statement it would be false: `if (!some_bool!) { some_code_here; }`. That is, instead of having to flip the boolean somewhere in the if-statement, it gets flipped automatically if the if-statement body executes (i.e. the check succeeds). So, in the above if-statement, if `some_bool` was false to begin with, after checking if it is true or not is not enough for it to flip, only if the check is true and the if-statement body executes would it flip.

To be concise, `if (!some_bool!) { some_code_here; }` would be equivalent to `if (some_bool) { some_bool = !some_bool; some_code_here; }`. The negation, of course, would be the same as normal, i.e. `if (!!some_bool!) { some_code_here; }`.

**Comment by jnm2** [comment]
Postfix `!` is currently allowed already as part of the Nullable Reference Types feature of C# 8, so I think this could change the meaning of existing valid code.

[SharpLab](https://sharplab.io/#v2:C4LgbgpgTglgZgTwD4AEBMACAwgWAFADe+GJGKALBgLIAUARgPYMA2GAxgwHYAmAlMaSJ5SIjPAw0AhBx6T+w0SSGKSAXwFr86vEA===)

```cs
class C
{
    void M(bool cond)
    {
        if (!cond!)
        {
        }
    }
}
```

**Reply by scott-parkhill** [reply]
Hmm, I see what you mean, are you referring to the [null-forgiving operator](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/operators/null-forgiving)? I had thought of this, but never would have thought it would occur in tandem with a logical negation operator. Perhaps this assumption is wrong? I thought there would be `expr!` and `!expr` but never `!expr!` currently. 

Edit: you're probably right though. Rip.

**Reply by jnm2** [reply]
There would have to have been something specifically added to block that composition of operators, and there wasn't.

**Comment by FaustVX** [comment]
The idea is good, but there are some issues.

- Like @jnm2 said, it's already a valid solution with NRT
- If you don't want to repeat the expression, you could write `expr ^= true`, the XOR `^` will negate the first operand only if the second is true
- the `if (!expr!) { }` is not symmetrical with the actual behavior :
  - Currently, the condition inside the `if` is executed before his body
    - and what happens with else : `if (!expr!) { } else { }` ?
  - if `!expr!` is lowered to `expr = !expr` then the body will be executed only if `expr` was false before

**Comment by Unknown6656** [comment]
I always use:
```csharp
value ^= true;
```
If you do not like this, you could use an extension method (note the `this ref`!):
```csharp
public static bool flip(this ref bool b) => b ^= true;

bool value = .....;
value.flip();
```
Feel free to rename `flip` to `f` to save characters 😉

**Reply by scott-parkhill** [reply]
That's nifty, thanks! Good idea.

**Comment by ronnygunawan** [comment]
```cs
bool b = true;
b--; // Postfix boolean negation
--b; // Prefix boolean negation

// or introduce a new !! operator

b!!;
!!b;
```

**Reply by jnm2** [reply]
`!!b` has an existing valid meaning. `b!!` does not though.

**Reply by TheUnlocked** [reply]
It's worth noting that `b!!` could be confusing to people coming from Kotlin since that's their not-null assertion operator (similar to C#'s null forgiving operator).

If we're trying to fit in with how C# already does things I almost feel like `b !=;` would be better, to match with the other `op=` operators.

**Reply by Joe4evr** [reply]
Also, `b!!` is (last I checked) the current proposed operator for making the compiler automatically insert a parameter null-check.

**Reply by scott-parkhill** [reply]
I like `b !=;` to be honest, it's clear and does the job well and I can't think of anything it could be confused with.

**Reply by CyrusNajmabadi** [reply]
that just seems not better than `b = !b;` :)

**Reply by jnm2** [reply]
Just don't forget that `b` is actually hiding `some.Possibly.Grotesquely.Long.Lvalue.Expression`
